frontends/p4/unusedDeclarations.h

```diff
@@ -20,19 +20,62 @@ limitations under the License.
 #include "../common/resolveReferences/resolveReferences.h"
 #include "ir/ir.h"
 #include "ir/pass_manager.h"
+#include "lib/iterator_range.h"
+#include "lib/stringify.h"
 
 namespace P4 {
 
 class RemoveUnusedDeclarations;
 
+class UsedDeclSet : public IHasDbPrint {
+    /// Set containing all declarations in the program.
+    absl::flat_hash_set<const IR::IDeclaration *, Util::Hash> usedDecls;
+
+ public:
+    bool setUsed(const IR::IDeclaration *decl) { return usedDecls.emplace(decl).second; }
+
+    [[nodiscard]] auto begin() const { return usedDecls.begin(); }
+
+    [[nodiscard]] auto end() const { return usedDecls.end(); }
+
+    [[nodiscard]] auto used() const { return Util::iterator_range(usedDecls); }
+
+    void clear() { usedDecls.clear(); }
+
+    void dbprint(std::ostream &cout) const override;
+
+    /// @returns @true if @p decl is used in the program.
+    bool isUsed(const IR::IDeclaration *decl) const { return usedDecls.count(decl) > 0; }
```

Comment:
I believe there is C++20 style `.contains` in the Abseil containers, it would be better to use that.